### PR TITLE
Feature/unique id&login

### PR DESCRIPTION
**Pull request by nklapste**
Adding unique Id and login validation.

Also hooked this functionality into the `LoginPresenter` and `LoginCareProviderActivity` and `LoginPatientActivity`.

These tests are still abit finicky as they rely on the fact that our testing ElasticSearch remote contains the users `TESTPatient1` and `TESTCareProvider1` both with the password `foobar123`

Removed checks for a unique username in  `Username`

### CODE REVIEW

File: app/src/main/java/ca/klapstein/baudit/models/DataModel.java [hunk 1]
```diff
@@ -47,14 +42,21 @@ static public boolean uniqueID(String username) {
...
-     * @param username {@code String}
-     * @param password {@code String}
+     * considering param type {@code String}
+     * @param username {@code Username}
+     * @param password {@code Password}
      * @return {@code boolean}
      */
-    public boolean validateLogin(String username, String password) {
-        // TODO: implement from RemoteModel
+    public boolean validateLogin(Username username, Password password) {
```
Comment (dcones) [inline app/src/main/java/ca/klapstein/baudit/models/DataModel.java]: Would accepting Username and Password classes increase coupling? As opposed to passing username.getUsernameString and password.getPassword during calls to validateLogin? I recognize though that it makes it a bit cleaner to retrieve the string in validateLogin.

File: app/src/main/java/ca/klapstein/baudit/models/DataModel.java [hunk 2]
```diff
@@ -35,10 +32,8 @@ public DataModel(Context context) {
      * @param username {@code String}
      * @return {@code true} if the username {@code String} representation is not already taken, otherwise {@code false}
      */
-    static public boolean uniqueID(String username) {
-        // TODO: validate from the local?
-        // TODO: implement from RemoteModel
-        return true;
+    public boolean uniqueID(Username username) {
+        return getPatient(username) == null && getCareProvider(username) == null;
```
Comment (dcones) [inline app/src/main/java/ca/klapstein/baudit/models/DataModel.java]: We are allowing a patient and a care provider to have the same username?

File: app/src/main/java/ca/klapstein/baudit/models/DataModel.java [hunk 3]
```diff
@@ -35,10 +32,8 @@ public DataModel(Context context) {
      * @param username {@code String}
      * @return {@code true} if the username {@code String} representation is not already taken, otherwise {@code false}
```
Comment (dcones) [inline app/src/main/java/ca/klapstein/baudit/models/DataModel.java]: @code Username

File: app/src/main/java/ca/klapstein/baudit/models/RemoteModel.java [hunk 4]
```diff
@@ -191,15 +265,14 @@ protected CareProvider doInBackground(String... search_parameters) {
...
-                            "     }\n" +
-                            "}";
+            String query = "{\n" +
+                    "    \"query\": {\n" +
+                    "        \"ids\" : {\n" +
+                    "            \"type\" : \"careprovider\",\n" +
+                    "            \"values\" : [\"" + search_parameters[0] + "\"]\n" +
+                    "         }\n" +
+                    "     }\n" +
+                    "}";
```
Comment (dcones) [inline app/src/main/java/ca/klapstein/baudit/models/RemoteModel.java]: Maybe we could pass an extra type parameter to reduce code duplication?

File: app/src/main/java/ca/klapstein/baudit/models/DataModel.java (hunk 2 above)
Comment (nklapste) [reply app/src/main/java/ca/klapstein/baudit/models/DataModel.java]: this checks that there exists no `Patient` with the given username

it also checks that there exists no `CareProvider` with the given username

either either a `CareProvider` or `Patient` exists with a username it is not a unique username

thus, two patient(s) and/or care provider(s) cannot share the same username.

File: app/src/main/java/ca/klapstein/baudit/models/RemoteModel.java (hunk 4 above)
Comment (nklapste) [reply app/src/main/java/ca/klapstein/baudit/models/RemoteModel.java]: maybe add as a TODO: for later refactoring

**Pull request merged**